File: src/backends/webgpu/src/conv2d_test.ts
/**
 * @license
 * Copyright 2019 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

import * as tf from '@tensorflow/tfjs-core';
import {expectArraysClose} from '@tensorflow/tfjs-core/dist/test_util';

import * as tfwebgpu from './index';

describe('WebGPU backend - convolution tests', () => {
  beforeAll(() => tfwebgpu.ready);

  it('x=[1,4,4,1] f=[1,1,1,3] s=2 d=1 p=same', async () => {
    const inputDepth = 1;
    const inputShape: [number, number, number] = [4, 4, inputDepth];
    const outputDepth = 3;
    const fSize = 1;
    const pad = 'same';
    const stride: [number, number] = [2, 2];

    const x = tf.tensor3d(
        [
          10, 30, 50, 70, 20, 40, 60, 80, -10, -30, -50, -70, -20, -40, -60, -80
        ],
        inputShape);
    const w = tf.tensor4d([1, 0.5, 1], [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 3]);
    expectArraysClose(
        resultData,
        new Float32Array([10, 5, 10, 50, 25, 50, -10, -5, -10, -50, -25, -50]));
  });

  it('x=[3,3,2] f=[2,2,2,1] s=1 d=1 p=valid', async () => {
    const pad = 'valid';
    const stride = 1;

    const x = tf.tensor3d(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20, 30, 40, 50, 60, 70, 80, 90],
        [3, 3, 2]);
    const w = tf.tensor4d([1, 2, 3, 4, 5, 6, 7, 8], [2, 2, 2, 1]);

    const result = tf.conv2d(x, w, stride, pad);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 1]);
    expectArraysClose(resultData, new Float32Array([
                        256,
                        535,
                        1570,
                        2209,
                      ]));
  });

  it('x=[1,3,6,1] f=[2,2,1,1] s=[1,2] d=1 p=valid', async () => {
    const inputDepth = 1;
    const inputShape: [number, number, number, number] = [1, 3, 6, inputDepth];
    const outputDepth = 1;
    const fSize = 2;
    const pad = 'valid';
    const stride: [number, number] = [1, 2];

    const input =
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18];
    const filter = [1, 2, 3, 4];

    const x = tf.tensor4d(input, inputShape);
    const w = tf.tensor4d(filter, [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad);

    const resultData = await result.data();
    expect(result.shape).toEqual([1, 2, 3, 1]);
    expectArraysClose(
        resultData, new Float32Array([58, 78, 98, 118, 138, 158]));
  });

  it('x=[2,2,1] f=[2,2,1,1] s=1 d=1 p=0', async () => {
    const inputDepth = 1;
    const inputShape: [number, number, number] = [2, 2, inputDepth];
    const outputDepth = 1;
    const fSize = 2;
    const pad = 0;
    const stride = 1;
    const dataFormat = 'NHWC';
    const dilation = 1;

    const x = tf.tensor3d([1, 2, 3, 4], inputShape);
    const w =
        tf.tensor4d([3, 1, 5, 0], [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([1, 1, 1]);
    expectArraysClose(resultData, new Float32Array([20]));
  });

  it('x=[2,2,2,1] f=[1,1,1,1] s=1 d=1 p=0', async () => {
    const inputDepth = 1;
    const inShape: [number, number, number, number] = [2, 2, 2, inputDepth];
    const outputDepth = 1;
    const fSize = 1;
    const pad = 0;
    const stride = 1;

    const x = tf.tensor4d([1, 2, 3, 4, 5, 6, 7, 8], inShape);
    const w = tf.tensor4d([2], [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad);
    expect(result.shape).toEqual([2, 2, 2, 1]);
    const expected = [2, 4, 6, 8, 10, 12, 14, 16];

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 2, 1]);
    expectArraysClose(resultData, new Float32Array(expected));
  });

  it('x=[2,2,1] f=[2,2,1,1] s=1 d=1 p=same', async () => {
    const inputDepth = 1;
    const inputShape: [number, number, number] = [2, 2, inputDepth];
    const outputDepth = 1;
    const fSize = 2;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NHWC';
    const dilation = 1;

    const x = tf.tensor3d([1, 2, 3, 4], inputShape);
    const w =
        tf.tensor4d([3, 1, 5, 0], [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 1]);
    expectArraysClose(resultData, new Float32Array([20, 26, 13, 12]));
  });

  it('x=[2,4,1] f=[4,2,1,1] s=1 d=1 p=same', async () => {
    const inputDepth = 1;
    const outputDepth = 1;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NHWC';
    const dilation = 1;

    const x = tf.tensor3d([1, 2, 3, 4, 5, 6, 7, 8], [2, 4, inputDepth]);
    const w =
        tf.tensor4d([3, 1, 5, 0, 2, 7, 8, 9], [4, 2, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 4, 1]);
    expectArraysClose(
        resultData, new Float32Array([57, 71, 85, 36, 30, 39, 48, 52]));
  });
});

File: tfjs-backend-wasm/src/cc/kernels/Conv2D_test.cc
/* Copyright 2019 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ===========================================================================*/

#include <gtest/gtest.h>

#include <xnnpack.h>
#include <cstddef>
#include <vector>

#include "src/cc/backend.h"
#include "src/cc/kernels/Conv2D.h"
#include "src/cc/util.h"

TEST(CONV2D, xnn_operator_lifetime) {
  tfjs::wasm::init();

  ASSERT_EQ(0, tfjs::backend::num_tensors());

  const size_t x0_id = 1;
  const size_t x1_id = 2;
  const size_t size = 8;
  float x_values[size] = {1, 2, 3, 4, 5, 6, 7, 8};

  const size_t weights0_id = 3;
  const size_t weights1_id = 4;
  const size_t weights_size = 8;
  float weights_values[weights_size] = {1, 2, 3, 4, 5, 6, 7, 8};

  const size_t out_id = 5;
  const size_t out_size = 12;
  float out_values[out_size] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

  tfjs::wasm::register_tensor(x0_id, size, x_values);
  tfjs::wasm::register_tensor(x1_id, size, x_values);
  tfjs::wasm::register_tensor(weights0_id, weights_size, weights_values);
  tfjs::wasm::register_tensor(weights1_id, weights_size, weights_values);
  tfjs::wasm::register_tensor(out_id, out_size, out_values);

  ASSERT_EQ(5, tfjs::backend::num_tensors());
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  // One xnn_operator should be created for the first call to conv2d.
  const size_t batch_size = 1;
  const size_t input_height = 4;
  const size_t input_width = 2;
  const size_t filter_height = 4;
  const size_t filter_width = 2;
  const size_t pad_top0 = 1;
  const size_t pad_right = 0;
  const size_t pad_bottom0 = 0;
  const size_t pad_left = 0;
  const bool is_same_pad0 = false;
  const size_t dilation_height = 1;
  const size_t dilation_width = 1;
  const size_t stride_height = 1;
  const size_t stride_width = 1;
  const size_t input_channels = 1;
  const size_t output_channels = 1;
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights0_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // No new xnn_operators should be created for the second call to conv2d with
  // the same arguments.
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights0_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // No new xnn_operators should be created for the second call to conv2d with
  // the same arguments but different input.
  tfjs::wasm::Conv2D(x1_id, batch_size, input_height, input_width, weights0_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // One new xnn_operator should be created for the next call to conv2d with the
  // same weights but different arguments.
  const size_t pad_top1 = 0;
  const size_t pad_bottom1 = 1;
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights0_id,
                     filter_height, filter_width, pad_top1, pad_right,
                     pad_bottom1, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // new weights and same input.
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights1_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(3, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // 'SAME' padding.
  const bool is_same_pad1 = true;
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights1_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad1, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(4, tfjs::backend::xnn_operator_count);

  // No new XNN operators should be created for the next call to conv2d with
  // 'SAME' padding and different input and raw padding values.
  tfjs::wasm::Conv2D(x1_id, batch_size, input_height, input_width, weights1_id,
                     filter_height, filter_width, pad_top1, pad_right,
                     pad_bottom1, pad_left, is_same_pad1, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(4, tfjs::backend::xnn_operator_count);

  // Disposing the first weights should remove 2 operators.
  tfjs::wasm::dispose_data(weights0_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // Disposing the second weights should remove the last 2 operator.
  tfjs::wasm::dispose_data(weights1_id);
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  tfjs::wasm::dispose();
}

TEST(CONV2D, transposed_filter_lifetime) {
  tfjs::wasm::init();

  ASSERT_EQ(0, tfjs::backend::num_tensors());

  const size_t x0_id = 1;
  const size_t weights1_id = 2;

  const size_t x_size = 8;
  const size_t weights_size = 8;
  std::vector<float> x{1, 2, 3, 4, 5, 6, 7, 8};
  std::vector<float> weights{1, 2, 3, 4, 5, 6, 7, 8};

  const size_t out_id = 3;
  const size_t out_size = 12;
  std::vector<float> out{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

  tfjs::wasm::register_tensor(x0_id, x_size, x.data());
  tfjs::wasm::register_tensor(weights1_id, weights_size, weights.data());
  tfjs::wasm::register_tensor(out_id, out_size, out.data());

  ASSERT_EQ(3, tfjs::backend::num_tensors());
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  // One xnn_operator should be created for the first call to conv2d.
  const size_t batch_size = 1;
  const size_t input_height = 4;
  const size_t input_width = 2;
  const size_t filter_height = 4;
  const size_t filter_width = 2;
  const size_t pad_top0 = 1;
  const size_t pad_right = 0;
  const size_t pad_bottom0 = 0;
  const size_t pad_left = 0;
  const bool is_same_pad0 = false;
  const size_t dilation_height = 1;
  const size_t dilation_width = 1;
  const size_t stride_height = 1;
  const size_t stride_width = 1;
  const size_t input_channels = 1;
  const size_t output_channels = 1;
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights1_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);
  EXPECT_EQ(out, (std::vector<float>{133, 204, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}));

  // Make a new filter in case the memory of the first filter was released and
  // this filter reuses it.
  std::vector<float> weights2 = {8, 7, 6, 5, 4, 3, 2, 1};
  const size_t weights2_id = 4;
  tfjs::wasm::register_tensor(weights2_id, weights_size, weights2.data());

  // No new xnn_operators should be created for the second call to conv2d with
  // the same filter.
  tfjs::wasm::Conv2D(x0_id, batch_size, input_height, input_width, weights1_id,
                     filter_height, filter_width, pad_top0, pad_right,
                     pad_bottom0, pad_left, is_same_pad0, dilation_height,
                     dilation_width, stride_height, stride_width,
                     input_channels, output_channels, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // The output should be the same as the first time since we are still using
  // the first filter.
  EXPECT_EQ(out, (std::vector<float>{133, 204, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}));

  tfjs::wasm::dispose_data(weights1_id);
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  tfjs::wasm::dispose();
}
